staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -360,6 +396,43 @@ func (r *crdHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 		)
 		return
 	}
+	// Make sure storage version is updated before we serve write requests
+	if utilfeature.DefaultFeatureGate.Enabled(features.StorageVersionAPI) &&
+		utilfeature.DefaultFeatureGate.Enabled(features.APIServerIdentity) &&
+		crdInfo.storageVersionUpdateProcessed != nil &&
+		(requestInfo.Verb == "create" ||
+			requestInfo.Verb == "update" ||
+			requestInfo.Verb == "patch") {
+		select {
+		case <-crdInfo.storageVersionUpdateProcessed:
+		case <-crdInfo.storageVersionUpdateAbandoned:
+			utilruntime.HandleError(fmt.Errorf("failed to update storage version for custom resource, group: %v, resource: %v",
+				requestInfo.APIGroup, requestInfo.Resource))
+			// Storage version update failed because the CRD was deleted. Here
+			// we return a 503 to fail this in-flight request and let the client
+			// retry. If the CRD gets re-created later, the client will be able
+			// to send new requests. Otherwise, the client will receive 404 when
+			// the CRD teardown is finished.
+			responsewriters.ErrorNegotiated(
+				apierrors.NewServiceUnavailable("storage version update failed"),
+				Codecs, schema.GroupVersion{Group: requestInfo.APIGroup, Version: requestInfo.APIVersion}, w, req,
+			)
+			return
+		// Unblock the requests if the storage version update takes a long time, otherwise
+		// CR requests may stack up and overwhelm the API server. We will benchmark the
+		// storage version update latency to adjust the timeout in future.
+		// NOTE: currently the graceful CRD deletion waits 1s for in-flight requests
+		// to register themselves to the wait group. Ideally the storage version update should
+		// not cause the requests to miss the 1s window; otherwise the requests may
+		// fail ungracefully (e.g. it may happen if the CRD was deleted immediately after the first CR request
+		// establishes the underlying storage).
```

Comment:
nit: Can you move this "NOTE" to the beginning of the `select`? Given that `storageVersionUpdateTimeout` is 15s now, this timeout case won't help guarantee the graceful shutdown anyway.